File: filethumbnails/js/js/scenario2.js
//// THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF
//// ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO
//// THE IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//// PARTICULAR PURPOSE.
////
//// Copyright (c) Microsoft Corporation. All rights reserved

(function () {
    "use strict";

    var page = WinJS.UI.Pages.define("/html/scenario2.html", {
        ready: function (element, options) {
            document.getElementById("music-thumb-start").addEventListener("click", pickMusic, false);
        }
    });

    function pickMusic() {
        // Set API parameters
        var requestedSize = 100,
            thumbnailMode = Windows.Storage.FileProperties.ThumbnailMode.musicView;

        // Clean output in case of repeat usage
        cleanOutput();

        var openpicker = new Windows.Storage.Pickers.FileOpenPicker();
        openpicker.fileTypeFilter.replaceAll([".mp3", ".wma", ".m4a", ".aac"]);
        openpicker.suggestedStartLocation = Windows.Storage.Pickers.PickerLocationId.musicLibrary;
        openpicker.pickSingleFileAsync().done(function (file) {
            if (file) {
                file.getThumbnailAsync(thumbnailMode, requestedSize).done(
                    function (thumbnail) {
                        if (thumbnail) {
                            // Also verify the type is ThumbnailType.image (album art) instead of ThumbnailType.icon
                            // (which may be returned as a fallback if the file does not provide album art)
                            if (thumbnail.type === Windows.Storage.FileProperties.ThumbnailType.image) {
                                outputResult(file, thumbnail, "ThumbnailMode.musicView", requestedSize);
                            } else {
                                WinJS.log && WinJS.log(SdkSample.errors.noAlbumArt, "sample", "status");
                            }
                        } else {
                            WinJS.log && WinJS.log(SdkSample.errors.noAlbumArt, "sample", "status");
                        }
                    },
                    function (error) {
                        WinJS.log && WinJS.log(SdkSample.errors.fail, "sample", "status");
                    }
                );
            } else {
                WinJS.log && WinJS.log(SdkSample.errors.cancel, "sample", "status");
            }
        });
    }

    function outputResult(item, thumbnailImage, thumbnailMode, requestedSize) {
        document.getElementById("music-thumb-imageHolder").src = URL.createObjectURL(thumbnailImage, { oneTimeOnly: true });
        // Close the thumbnail stream once the image is loaded
        document.getElementById("music-thumb-imageHolder").onload = function () {
            thumbnailImage.close();
        };
        document.getElementById("music-thumb-modeName").innerText = thumbnailMode;
        document.getElementById("music-thumb-fileName").innerText = "File used: " + item.name;
        document.getElementById("music-thumb-requestedSize").innerText = "Requested size: " + requestedSize;
        document.getElementById("music-thumb-returnedSize").innerText = "Returned size: " + thumbnailImage.originalWidth + "x" + thumbnailImage.originalHeight;
    }

    function cleanOutput() {
        WinJS.log && WinJS.log("", "sample", "status");
        document.getElementById("music-thumb-imageHolder").src = "/images/placeholder-sdk.png";
        document.getElementById("music-thumb-modeName").innerText = "";
        document.getElementById("music-thumb-fileName").innerText = "";
        document.getElementById("music-thumb-requestedSize").innerText = "";
        document.getElementById("music-thumb-returnedSize").innerText = "";
    }
})();

File: Samples/FileThumbnails/cppwinrt/Scenario2.cpp
//*********************************************************
//
// Copyright (c) Microsoft. All rights reserved.
// This code is licensed under the MIT License (MIT).
// THIS CODE IS PROVIDED *AS IS* WITHOUT WARRANTY OF
// ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING ANY
// IMPLIED WARRANTIES OF FITNESS FOR A PARTICULAR
// PURPOSE, MERCHANTABILITY, OR NON-INFRINGEMENT.
//
//*********************************************************

#include "pch.h"
#include "Scenario2.h"
#include "Scenario2.g.cpp"
#include "SampleConfiguration.h"

using namespace winrt;
using namespace Windows::Foundation;
using namespace Windows::Storage;
using namespace Windows::Storage::FileProperties;
using namespace Windows::Storage::Pickers;
using namespace Windows::UI::Xaml;
using namespace Windows::UI::Xaml::Navigation;

namespace winrt::SDKTemplate::implementation
{
    Scenario2::Scenario2()
    {
        InitializeComponent();
    }

    void Scenario2::OnNavigatedTo(NavigationEventArgs const&)
    {
        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());
    }

    fire_and_forget Scenario2::GetThumbnailButton_Click(IInspectable const&, RoutedEventArgs const&)
    {
        auto lifetime = get_strong();

        Helpers::ResetOutput(rootPage, ThumbnailImage(), OutputTextBlock());

        // Pick a music file
        FileOpenPicker openPicker;
        openPicker.FileTypeFilter().ReplaceAll(FileExtensions::Music());
        StorageFile file = co_await openPicker.PickSingleFileAsync();

        if (file != nullptr)
        {
            const ThumbnailMode thumbnailMode = ThumbnailMode::MusicView;
            const uint32_t size = 100;
            StorageItemThumbnail thumbnail = co_await file.GetThumbnailAsync(thumbnailMode, size);

            // Also verify the type is ThumbnailType::Image (album art) instead of ThumbnailType::Icon
            // (which may be returned as a fallback if the file does not provide album art)
            if (thumbnail != nullptr && thumbnail.Type() == ThumbnailType::Image)
            {
                // Display the thumbnail
                Helpers::DisplayResult(ThumbnailImage(), OutputTextBlock(), thumbnailMode, size, file, thumbnail, false);
            }
            else
            {
                rootPage.NotifyUser(Errors::NoAlbumArt, NotifyType::StatusMessage);
            }
            if (thumbnail != nullptr)
            {
                thumbnail.Close();
            }
        }
        else
        {
            rootPage.NotifyUser(Errors::Cancel, NotifyType::StatusMessage);
        }
    }
}
